pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +187,629 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+		ErrOut: writer{klog.Error},
+		DryRun: false,
+	}
 	return nil
 }
 
-func (r *ReconcileNodeMaintenance) initMaintenanceStatus(nm *kubevirtv1alpha1.NodeMaintenance) error {
-	if nm.Status.Phase == "" {
-		nm.Status.Phase = kubevirtv1alpha1.MaintenanceRunning
-		pendingList, errlist := r.drainer.GetPodsForDeletion(nm.Spec.NodeName)
-		if errlist != nil {
-			return fmt.Errorf("Failed to get pods for eviction while initializing status")
+func (r *ReconcileNodeMaintenance) runCordonOrUncordon(cordonOn bool, dcheck DeadlineCheck) error {
+
+	countTaint, numDesiredTaints := CountDesiredTaintOnNode(r.node)
+	if (cordonOn && countTaint != numDesiredTaints) || (!cordonOn && countTaint != 0) {
+		if !dcheck.isExpired() {
+			if err := AddOrRemoveTaint(r.drainer.Client, r.node, cordonOn); err != nil {
+				r.reqLogger.Errorf("failed to set tain (action-set: %t) error: %v", cordonOn, err)
+				return err
+			}
+			r.reqLogger.Debugf("completed add/remove taint set=%t", cordonOn)
+		} else {
+			r.reqLogger.Info("skipped remove taint; time expired taintOn=", cordonOn)
 		}
-		if pendingList != nil {
-			nm.Status.PendingPods = GetPodNameList(pendingList.Pods())
+
+	} else {
+		r.reqLogger.Debugf("no change of taint required. set=%t", cordonOn)
+	}
+
+	if r.node.Spec.Unschedulable != cordonOn {
+		if !dcheck.isExpired() {
+                        /*
+			   r.node.Spec.Unschedulable = cordonOn
+			   if err := r.updateNode(r.node); err != nil {
+			           r.reqLogger.Errorf("failed to cordon. action: %t %v", cordonOn, err)
+			           return err
+			   }
+                        */
+			if err := drain.RunCordonOrUncordon(r.drainer, r.node, cordonOn); err != nil {
+				r.reqLogger.Errorf("failed to cordon. action: %t %v", cordonOn, err)
+				return err
+			}
+			r.reqLogger.Debugf("completed cordon/uncordon action-set=%t", cordonOn)
+		} else {
+			r.reqLogger.Info("skipped cordon/uncordon; time expired cordonOn=", cordonOn)
 		}
-		nm.Status.EvictionPods = len(nm.Status.PendingPods)
+	} else {
+		r.reqLogger.Debugf("no change of cordon required. set=%t", cordonOn)
+	}
 
-		podlist, err := r.drainer.Client.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{
-			FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": nm.Spec.NodeName}).String()})
-		if err != nil {
+	return nil
+}
+
+func (r *ReconcileNodeMaintenance) evictPods(dcheck DeadlineCheck) error {
+
+	if !dcheck.isExpired() {
+		nodeName := r.node.ObjectMeta.Name
+		list, errs := r.drainer.GetPodsForDeletion(nodeName)
+		if errs != nil {
+			err := utilerrors.NewAggregate(errs)
+			r.reqLogger.Errorf("failed got get pod for eviction %v", err)
 			return err
 		}
-		nm.Status.TotalPods = len(podlist.Items)
-		err = r.client.Status().Update(context.TODO(), nm)
+
+		if !dcheck.isExpired() && len(list.Pods()) != 0 {
+
+			if dcheck.isSet && r.drainer.Timeout > dcheck.DurationUntilExpiration() {
+				r.drainer.Timeout = dcheck.DurationUntilExpiration()
```

Comment:
If this is our first time through the `Reconcile()` loop for a 3h maintenance window, the line above tells the system to wait up to the entire 3 hour period before processing any other maintenance requests.

The most we should wait for any eviction, deletion, or cordon request is:
``EvictionTimeSlice           = 10 * time.Second // max. time in seconds that eviction may block.``